torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp

```diff
@@ -1292,26 +1283,38 @@ void initNvFuserPythonBindings(PyObject* module) {
   nvf_ops.def(
       "batch_norm",
       [](nvfuser::FusionDefinition::Operators& self,
-         nvfuser::Tensor x,
-         nvfuser::Tensor weight,
-         nvfuser::Tensor bias,
-         nvfuser::Tensor running_mean,
-         nvfuser::Tensor running_var,
-         bool training,
+         nvfuser::Tensor arg,
+         c10::optional<nvfuser::Tensor> weight,
+         c10::optional<nvfuser::Tensor> bias,
+         c10::optional<nvfuser::Tensor> running_mean,
+         c10::optional<nvfuser::Tensor> running_var,
          nvfuser::Scalar momentum,
          nvfuser::Scalar eps,
+         bool training,
```

Comment:
I have a mixed feeling here. Why do we not want to keep the same arg order as with `aten::batch_norm`?

Reply:
The reason I had to re-order the params is because `training` and `channels_last` were not defined as `State`.  `State` objects are mechanically added in front of any Fusion Record specific attributes.  I chose to do the simplest, thing.  That was to re-order the parameters.  All the choices:

1. Reorder parameters `State` in front of the Fusion Record Attributes.
2. Make an explicit list of the list of items to be printed to allow for intermixing.
3. Turn `training` and `channels_last` into state.

Reply:
> `State` objects are mechanically added in front of any Fusion Record specific attributes.

But that's in fusion record right?

For exposed python binding, I thought the order of arguments doesn't affect how we construct IRs in fusion record and this should be just a mechanical change.

Reply:
It impacts the printing. The printing mechanically goes through the `args` first.  So, if you try to re-use the printed function and `training` is in the middle of the `State` args, it will fail.

It was a choice not to make `training` and `channels_last` state args.

https://github.com/pytorch/pytorch/blob/139afc50ecafcde5fb085e2cca78fed55e6b5aad/torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h#L1451

https://github.com/pytorch/pytorch/blob/139afc50ecafcde5fb085e2cca78fed55e6b5aad/torch/csrc/jit/codegen/cuda/python_frontend/fusion_record.h#L140-L154

Reply:
Sorry if I'm trying too hard with nitpicking :stuck_out_tongue: 

That requirement seems to only apply to `RecordFunctor` though.

I was suggesting that our python API to match aten signature. `nvf_ops.def("batch_norm", ..., bool training, nvfuser::Scalar momentum, nvfuser::Scalar eps)`
But later in the record definition, we can still do
```
        fd->defineRecord(new nvfuser::BatchNormOpRecord(
            {fd->recordingState(arg()),
             weight_state,
             bias_state,
             running_mean_state,
             running_var_state,
             fd->recordingState(momentum()),
             fd->recordingState(eps())},
            {fd->recordingState(output()),
             fd->recordingState(mean()),
             fd->recordingState(invstd())},
            training,
            channels_last));
```